rfd/0029-account-lifecycle.md

```diff
@@ -0,0 +1,220 @@
+---
+authors: Sasha Klizhentas (sasha@goteleport.com)
+state: draft
+---
+
+# RFD 29 - Account Lifecycle: Recovery and Cancellation
+
+## What
+
+Introduces cloud account recovery and cancellation flow.
+
+## Why
+
+Having a secure self-management account recovery and cancelation
+flow improves user experience and reduces support load on teams.
+
+## Details
+
+The recovery flow focuses on local accounts.
+SSO account recovery is covered by "All is lost" scenario.
+
+The flow will work for cloud and non-cloud accounts. For cloud it should be impossible
+to turn off second factor. For self-hosted, if second factor is off, the account recovery
+should not be possible.
+
+Each local user will receive 3 account recovery tokens on signup.
+
+Each token is a crypto-randomly generated passphrase string that could be used only once.
+
+We will use the vendored EFF long list via diceware library
+
+https://github.com/sethvargo/go-diceware
+
+The token is 8 word token resulting in roughly [~100 bits of entropy](https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases).
+
+Once signed up, users will be presented with the recovery tokens printed on the
+screen with a message:
+
+```
+Please save these account recovery tokens in a safe offline place.
+You can use each token once if you loose your second factor or a password.
+
+* tele-example-teleport-sh-cosponsor-happening-moocher-undrilled-royal-depravity-frustrate-starting
+* tele-example-teleport-sh-factoid-shorten-koala-explore-retinal-lazy-coauthor-brethren
+* tele-example-teleport-sh-thinness-suspect-shrink-dwindle-awning-engulf-thrift-spoiling
+```
+
+Customers can use cloud prefix in case if they forget the domain name as well.
+Teleport cloud scanners can look for `tele-` prefix and alert customers with
+domain names in the prefix about leaked tokens.
+
+All 3 tokens are regenerated in batch every time account recovery is successfull.
+Each token can only be used one time successfully or unsuccessfully.
+
+Teleport should store tokens encrypted with `crypto/bcrypt` and compare using constant-time compare.
```

Comment:
I think considerations are probably needed around rate limiting the API as well.